### DISCUSSION TITLE
第三方库使用的方法求指点

### DISCUSSION THREAD

**Discussion by nomadfighter** [xmake-io/xmake#1653] [Q&A]
由于各种原因，我希望使用远程包，但是又不想弄私有仓库，所以我写的一个包是这样子的

    package("physfs")
        add_urls("https://github.com/icculus/physfs.git")
        add_versions("github", "23690f316d33f04c9f9910f6ff1c9e5cb876db07")
    
        add_deps("cmake")
    
        on_install(function(package)
            import("package.tools.cmake").install(package, 
                {"-DPHYSFS_BUILD_STATIC=ON", "-DPHYSFS_BUILD_SHARED=OFF", "-DPHYSFS_BUILD_TEST=OFF"}
            )
        end)

xmake.lua代码如下

      add_rules("mode.debug", "mode.release")
      
      add_repositories("3rd 3rd")
      
      --这里的困扰特别大，如果不指定main，就会默认master，但是physfs的git仓库又没有master分支，导致各种错误。
      --指定了branch是main，前面package里面指定的commit id似乎无法生效
      --需求是能clone指定commit id版本的代码，而不是main分支的最新代码
      add_requires("physfs main")  
      
      target("foo")
          set_kind("binary")
          add_files("src/*.cpp")
          
          add_packages("physfs")

最后运行xmake build -y
直接link error，应该是未能正确指定链接库，但是include路径似乎ok，所以问题到底是出在哪里？求教。
另外，如何使clone的代码放到本地目录，比如在.xmake目录或者build目录下，而不是跑到AppData/Local/.xmake下面。

总体上写的过程比cmake舒服太多，lua代码很好看，也相对严谨，不像cmake有些地方实在恶心。
但是资料确实有点少，google搜索基本只能搜出来什么xmake vs cmake之类的东西，具体的使用示例
几乎找不到，要不上来就是一个很大的工程。要不就是官方文档那样，给个命令说明但是又没啥上下文，比如
这个package，我的本意是想看看“如何使用第三方package”，但是官方文档上来就让我运行xmake package，

**Comment by waruqi** [comment]
>   add_versions("github", "23690f316d33f04c9f9910f6ff1c9e5cb876db07")

可以指定commiy id，但是前面必须是规范的语义版本号，比如 x.x.x，如果是git commit 你可以用commit date作为版本号比如 2021.9.6 这样才能被自动作为 latest version选择到

> 直接link error，应该是未能正确指定链接库，但是include路径似乎ok，所以问题到底是出在哪里？求教。

所以最好加个 on_test 测试验证下，这样可以在包安装阶段就提前报错，而不是编译项目的时候。

默认 xmake 会自动帮你找所有安装的 links ，但是也有例外，具体看每个包自身的情况，如果没找到 links，自己在包定义里面 add_links 手动配置上

你可以去 https://github.com/xmake-io/xmake-repo
仓库随便找几个包看看，参考下，或者加 -vD 看详情

> 这个package，我的本意是想看看“如何使用第三方package”，但是官方文档上来就让我运行xmake package，

文档有关于包的详细说明，你没找对地方，包括如何做包

https://xmake.io/#/zh-cn/package/remote_package?id=%e6%b7%bb%e5%8a%a0%e5%8c%85%e5%88%b0%e4%bb%93%e5%ba%93

**Reply by waruqi** [reply]
之后到这里讨论

**Reply by nomadfighter** [reply]
感谢作者回复，
我就是参考
https://github.com/xmake-io/xmake-repo/blob/7d6edbc088e18bba8f413d9bee16d932b1f0153e/packages/b/bgfx/xmake.lua#L8
这里得写法，他这个7816感觉很随意得样子，我翻了好久bgfx的仓库，也没见着有7816这个版本号
然后我本地测试这个一直提示
=> clone https://github.com/bkaradzic/bgfx.git 7816 .. failed

**Reply by waruqi** [reply]
他那个版本格式也不对 回头我修下  你看下这个

https://github.com/xmake-io/xmake-repo/blob/38f26a304189e28067903eddc77f53bf2978181e/packages/f/fastcppcsvparser/xmake.lua#L7

**Reply by xq114** [reply]
> 感谢作者回复，
> 我就是参考
> [xmake-io/xmake-repo@`7d6edbc`/packages/b/bgfx/xmake.lua#L8](https://github.com/xmake-io/xmake-repo/blob/7d6edbc088e18bba8f413d9bee16d932b1f0153e/packages/b/bgfx/xmake.lua#L8)
> 这里得写法，他这个7816感觉很随意得样子，我翻了好久bgfx的仓库，也没见着有7816这个版本号
> 然后我本地测试这个一直提示
> => clone [bkaradzic/bgfx.git](https://github.com/bkaradzic/bgfx.git) 7816 .. failed

https://github.com/bkaradzic/bgfx/blob/master/src/version.h 这里有版本。7816是一个版本号（仅有大版本号，符合semantic version的）

**Reply by waruqi** [reply]
我这边本地测试 bgfx 是可以过得么，是不是你那 xmake 不是最新版，我记得之前的版本 对 select version 上有点问题，后来修过。

**Reply by nomadfighter** [reply]
xmake v2.5.7+202108290043, A cross-platform build utility based on Lua
操作系统win10  19042

还是说我的代码不对头？

        add_rules("mode.debug", "mode.release")
        
        add_requires("bgfx master")
        
        target("foo")
            set_kind("binary")
            add_files("src/*.cpp")
            add_packages("bgfx")

看了下AppData\Local\.xmake\cache\packages目录，只有bgfx的代码，没有bx和bimg代码
所以install过程失败了

    => install bgfx master .. failed
    if you want to get verbose errors, please see:
    -> AppData\Local\.xmake\cache\packages\2109\b\bgfx\master\installdir.failed\logs\install.txt
    error: install failed!

这个install.txt的内容

      attempt to concatenate local 'p' (a nil value)

**Reply by waruqi** [reply]
>     add_requires("bgfx master")

你写死了 master 分支，但是这个包你看下它的定义，只有 7816 版本，才加了 bx/bimg 资源

要么直接  `add_requires("bgfx")` 自动获取最近版本，也就是， 7816 。。要么显式设置 `add_requires("bgfx 7816")`  要么改进 gbfx 包，对其他分支包括 master 也去拉取对应版本的 bx/img  资源。。

https://github.com/xmake-io/xmake-repo/blob/7d6edbc088e18bba8f413d9bee16d932b1f0153e/packages/b/bgfx/xmake.lua#L10-L11

**Reply by nomadfighter** [reply]
现在问题是不指定master的话，会提示
=> clone https://github.com/bkaradzic/bgfx.git 7816 .. failed
看起来似乎是要clone该仓库的“7816”分支，这显然不对头。
这里我还是没有理解，对于要使用git的包来说，
add_requires("bgfx xxx")后面这个xxx到底是指branch名字，还是版本号？
如果是branch，那7816显然不对，
如果是版本号，那这个版本号应该对应git仓库的什么内容？看前面
@xq114 的回复，这个版本号是从代码中取出来的，而不是其git仓库指定的版本号，
个人以为应该不会从源代码中搜索出版本号来使用吧？否则那速度不可能快起来啊

**Reply by waruqi** [reply]
```lua
add_requires("bgfx 7816")
```

```console
in xmake-repo:
  -> bgfx 7816
please input: y (y/n/m)
checking for ping ... /sbin/ping
pinging for the host(github.com) ... 107 ms
/usr/bin/git clone https://github.com/bkaradzic/bgfx.git source.tmp/bgfx
Cloning into 'source.tmp/bgfx'...
remote: Enumerating objects: 97972, done.
remote: Counting objects: 100% (525/525), done.
remote: Compressing objects: 100% (315/315), done.
remote: Total 97972 (delta 212), reused 448 (delta 201), pack-reused 97447
Receiving objects: 100% (97972/97972), 235.33 MiB | 15.53 MiB/s, done.
Resolving deltas: 100% (73271/73271), done.
Updating files: 100% (4539/4539), done.
/usr/bin/git checkout 5ecddbf4d51e2dda2a56ae8cafef4810e3a45d87
Note: switching to '5ecddbf4d51e2dda2a56ae8cafef4810e3a45d87'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 5ecddbf4d Updated version number.
  => clone https://github.com/bkaradzic/bgfx.git 7816 .. ok
```

这不可以的么，会优先走 semver 版本是否为有效语义版本，如果是，走 add_versions 里面的版本选择，如果不是，那尝试走 branch/tag 进行 checkout 

```console
$ xmake l core.base.semver.new 7816
7816
```

我看目前能正常解析作为语义版本。

**Reply by waruqi** [reply]
可以走 xmake -vD 看向详情，为啥clone failed，都走到 clone 了，说明版本选择都过掉了，没啥问题，都要开始 clone + checkout 了。。只是失败了，要么是网络问题

**Comment by nomadfighter** [comment]
error: @programdir\core\sandbox\modules\os.lua:98: cannot move source.tmp to source Permission denied

权限不够。。。跟我是域用户有关吗？我找找看是不是哪里有做了什么策略

**Reply by waruqi** [reply]
相关包安装的 cache 目录可能被占用也会出现这个问题，可以看下有没有其他终端在执行 xmake 访问这个目录，或者在执行安装操作。。

%localappdata%/.xmake/cache/ 下 b/bfgx/source 目录也可以手动删了试试

**Reply by nomadfighter** [reply]
这个问题已经解决，谢谢各位。
但是还有另一个问题
对于不使用git的package，通过curl下载会提示如下错误
curl: (3) URL using bad/illegal format or missing URL

但是我直接使用自己单独安装的curl下载没问题
curl -SL -x  -A "Xmake/2.5.7+202108290043 (Windows;) curl/7.78.0" https://github.com/fmtlib/fmt/releases/download/8.0.1/fmt-8.0.1.zip -o fmt-8.0.1.zip

查看xmake附带curl版本如下
curl 7.78.0 (i386-pc-win32) libcurl/7.78.0 OpenSSL/1.1.1l (Schannel) zlib/1.2.11 brotli/1.0.9 zstd/1.5.0 libidn2/2.3.2 libssh2/1.9.0 nghttp2/1.44.0 libgsasl/1.10.0
Release-Date: 2021-07-21
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli gsasl HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM SPNEGO SSL SSPI TLS-SRP UnixSockets zstd

自己安装的curl版本如下
curl 7.55.1 (Windows) libcurl/7.55.1 WinSSL
Release-Date: 2017-11-14, security patched: 2019-11-05
Protocols: dict file ftp ftps http https imap imaps pop3 pop3s smtp smtps telnet tftp
Features: AsynchDNS IPv6 Largefile SSPI Kerberos SPNEGO NTLM SSL

**Reply by waruqi** [reply]
这边也是从 https://curl.se/download.html 下载的最新版本 curl 。。你可以下载 7.78 版本手动试试，看看那个 版本可以

**Comment by myuanz** [comment]
我这几天一直在尝试封装包，文档翻来覆去地读，我十分确信“知识的诅咒”的存在。我希望我能搞懂然后写一个从零开始的教程
